Adding error raising tests

ludwiglierhammer opened the pull request:
Task
-----
cover more code with tests

ludwiglierhammer commented:
@jjk-code-otter: I created a new PR for adding more tests. Could you technically participate in this PR?

ludwiglierhammer commented on marine_qc/time_control.py:
```diff
@@ -491,8 +512,7 @@ def dayinyear(year: int, month: int, day: int) -> int:
 
-    if not (1 <= day <= 366):
-        raise ValueError(f"Invalid day in year: {day}. Must be between 1 and 366.")
+    assert 1 <= day <= 366
```
@jjk-code-otter: I think, we shold raise an error here, instead of using assert.

jjk-code-otter replied:
We have a choice. We can either assert and have complete test coverage, or we can have it raise an error if an impossible situation arises and not have complete test coverage, or we can delete the test.

There are a couple of situations like this in the code, where I can't think how to trigger the error in a unit test (easily). This happens because there are prior tests that stop improper values being processed and therefore we are guaranteed a valid output. Some of these were originally assertions because we were taught to use assertions to test inputs and outputs on some course.

What do you think? Should we keep them at the expense of complete test coverage?

jjk-code-otter replied:
I don't have a strong feeling either way. I'm happy to remove the assert statements. Is there someway to mark them so the coverage test doesn't flag them?

The advantage of assert rather than if...raise is that we can turn off assert when running and gain a little bump in performance. Most of our tests are just test-fail conditions. The exceptions aren't being caught, so the code just ends at that point. When we run with the CDM formatted data, lots of the checking will be unnecessary but it will happen anyway and there's no way to deactivate it.

ludwiglierhammer commented on marine_qc/time_control.py:
```diff
@@ -270,6 +270,35 @@ def pentad_to_month_day(p: int) -> tuple[int, int]:
     return m[p - 1], d[p - 1]
 
 
+def valid_month_day(month: int, day: int) -> bool:
```
@jjk-code-otter: I think, it would be helpfl if this routine is more flexible in a way that both `month` and `day` could be `None`. E.g. if `month` is `None`, `month_lengths` is `366.`

jjk-code-otter replied:
I'm not sure I understand this. In what situation would you want the month to be None?

jjk-code-otter replied:
I changed it to take either 2 arguments (month, day) or 3 arguments (year, month and day) the same as day_in_year.

ludwiglierhammer commented on marine_qc/time_control.py:
```diff
@@ -270,6 +270,35 @@ def pentad_to_month_day(p: int) -> tuple[int, int]:
     return m[p - 1], d[p - 1]
 
 
+def valid_month_day(month: int, day: int) -> bool:
```
In gerenal, I like the idea of this helper function.

ludwiglierhammer commented:
In  `time_control`, why do we have `da_in_year` and `dayinyear`? Why not combine them?

ludwiglierhammer commented on marine_qc/time_control.py:
```diff
@@ -491,8 +512,7 @@ def dayinyear(year: int, month: int, day: int) -> int:
 
-    if not (1 <= day <= 366):
-        raise ValueError(f"Invalid day in year: {day}. Must be between 1 and 366.")
+    assert 1 <= day <= 366
```
```suggestion
    assert 1 <= day <= 366  # noqa: S101
```

ludwiglierhammer commented on marine_qc/qc_grouped_reports.py:
```diff
@@ -168,8 +168,9 @@ def add_single_observation(
             raise ValueError(f"Invalid yindex: {yindex}. Must be between 0 and 180.")
-        if not (0 <= pindex < 73):
-            raise ValueError(f"Invalid pindex: {pindex}. Must be between 0 and 72.")
+        # pindex will be within this range or raise an Error from which pentad
+        # if not (0 <= pindex < 73):
+        #     raise ValueError(f"Invalid pindex: {pindex}. Must be between 0 and 72.")
```
Can we delete those lines?

jjk-code-otter commented on marine_qc/time_control.py:
```diff
@@ -295,58 +335,63 @@ def which_pentad(month: int, day: int) -> int:
-    Find the day number of a particular day from Jan 1st which is 1 to Dec 31st which is 365.
-
-    29 February is treated as 1 March
+def day_in_year(*args) -> int:
```
See e1a8815

ludwiglierhammer commented:
> I ran an additional coverage check, turning off `test_qc_on_db.py` as this calls lots of QC functions and was providing test coverage for some functions without actually testing them in detail (integration tests of a sort rather than unit tests). On turning off `test_qc_on_db.py` test coverage dropped quite a lot, but it's now back up to a reasonable level (around 94%) however there are still some obvious gaps, particularly in the following. I've added issues as noted or linked to existing issues:
> 
> 1. `multiple_row_checks.py` #14
> 2. `Climatology.py` related issues: #2
> 3. `external_clim.py` #17
> 4. `buoy_tracking_qc.py` #16
> 
> I've added these as separate issues because they don't pertain specifically to the intent of this pull request and are rather self-contained bits of work.

Thanks for this new issues. They are very helpful. Those are for future PRs.

jjk-code-otter commented:
> @jjk-code-otter: Is this PR ready to merge? If so, have you some suggestions for the CHANGELOG? I'll update it and solve the pre-commit issues.

Yes. Ready to merge. Thanks.

jjk-code-otter commented:
I made a small change in the CHANGELOG but I'm not sure what more to put.
